### PR TITLE
jzarr read/write (close #36)

### LINKED ISSUE #36: jzarr (v2)
cc: @grlee77 @SabineEmbacher

see https://github.com/zarr-developers/zarr_implementations/compare/master...joshmoore:jzarr?expand=1

Note: an upcoming version of jzarr will support nested and flat representations. The current code only generates default, flat arrays.

### PR DESCRIPTION
**Pull request by joshmoore**
The only real issue (repeated from #25) is that I don't see a way to return an array from the `read_from_jzarr` method which means that non-Python implementations may need to handle the verification code themselves internally.

cc: @constantinpape

### CODE REVIEW

File: generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        if (args.length != 0 && args.length != 3) {
+            System.out.println("usage: App");
+            System.out.println("usage: App -verify fpath dsname");
+            System.exit(2);  // EARLY EXIT
+        } else if (args.length == 3) {
+            String fpath = args[1];
+            String dsname = args[2];
+            ZarrArray verification = ZarrGroup.open(fpath).openArray(dsname);
+            int[] shape = verification.getShape();
+            if (!Arrays.equals(new int[]{}, shape)) {
```
Comment (grlee77) [inline generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java:81]: Maybe I am reading this wrong, but why does it look like this line is comparing a new empty array to shape? I was expecting `Array.equals(SHAPE, shape)`? (I am not a Java dev)

Am I correct that when called with arguments, this program only checks the shape, but not the actual image contents?

What do you think about writing out a NumPy `.npy` or `.npz` file from Java? (A quick search indicates there [is at least one library available for this](https://stackoverflow.com/questions/18727600/writing-npy-numpy-binary-format-from-java) ([GitHub](https://github.com/JetBrains-Research/npy)). That is what the xtensor-zarr executable does (using an .npz writer that was already available in `xtensor-io`). The array comparison to the reference can then done in Python. Or maybe it is just easier to do the verification in java itself instead?

Comment (joshmoore) [reply generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java:81]: > Maybe I am reading this wrong, but why does it look like this line is comparing a new empty array to shape? I was expecting `Array.equals(SHAPE, shape)`? (I am not a Java dev)

_Argh_. No, A) you're completely right. I added this before the troubles with `<u1` began in order to force a failure so B) why isn't it forcing a failure?!



> What do you think about writing out a NumPy `.npy` or `.npz` file from Java?

I'm personally skeptical. I think we're going to want to increase the amount of verification done in each language and it seems like that will hit limits via NumPy. But I assume I'm going to lose this one to you and @constantinpape 🙂

Comment (joshmoore) [reply generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java]: see: 7883082fdc38313d61a1b4e6940bace3bdcf158c

Comment (constantinpape) [reply generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java]: > I'm personally skeptical. I think we're going to want to increase the amount of verification done in each language and it seems like that will hit limits via NumPy. But I assume I'm going to lose this one to you and @constantinpape slightly_smiling_face

I am pretty agnostic on this one. I think it's fine if you validate in java. But we should somehow return the correct error codes to the python test to populate the test summary. (I haven't checked the rest of this PR yet, so maybe you are doing that already ;)).

Comment (grlee77) [reply generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java]: > why isn't it forcing a failure?!

I opened https://github.com/joshmoore/zarr_implementations/pull/1 with a simplified way of calling the subprocess that I am sure is actually running it (it takes a few seconds per test case). That one will fail if the exit code is not zero.

Comment (grlee77) [reply generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java]: I also do not have a problem with just doing the validation in Java

File: test/test_read_all.py [hunk 2]
```diff
@@ -43,6 +43,12 @@
 
 
 READABLE_CODECS: Dict[str, Dict[str, List[str]]] = {
+    "jzarr": {
+        "jzarr": ["blosc", "raw", "zlib"],
+        "zarr": ["blosc", "raw", "zlib"],
+        "xtensor_zarr": ["blosc", "raw", "zlib"],
+        "z5py": ["blosc", "raw", "zlib"],
+    },
```
Comment (grlee77) [inline test/test_read_all.py:51]:
```suggestion
    "jzarr": {
        "zarr": ["blosc", "raw", "zlib"],
        "zarr-v3": [],
        "N5": [],
    },
```

The keys here are the file format types, not the other library name (for `zarr` these two are the same so I think it happened to work okay as-is!).

File: generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            String fpath = args[1];
+            String dsname = args[2];
+            ZarrArray verification = ZarrGroup.open(fpath).openArray(dsname);
+            int[] shape = verification.getShape();
+            if (!Arrays.equals(SHAPE, shape)) {
+                throw new RuntimeException(String.format(
+                            "shape-mismatch expected:%s found:%s",
+                            Arrays.toString(SHAPE), Arrays.toString(shape)
+                ));
+            }
```
Comment (grlee77) [inline generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java:86]: It looks like the test verifies matching shape. Can `Arrays.equals` also be used to compare the values in `verification` to the values returned by `getData()`?

Comment (joshmoore) [reply generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java:86]: Done. Thanks for keeping me honest.

**Comment by joshmoore** [comment]
Looking at one of these failures (reading `js.zr/blosc/lz4/` with Zarr) the first few values differ:
Screen Shot 2021-05-04 at 21 37 17
Screen Shot 2021-05-04 at 21 37 26
Modifying `|u1` back to jzarr's `<u1` _doesn't_ correct the issue.

cc: @SabineEmbacher

**Comment by grlee77** [comment]
> Looking at one of these failures (reading js.zr/blosc/lz4/ with Zarr) the first few values differ:

Yeah, the values seen there are what would be obtained when trying to convert `uint8` values to `int8` so that values >=128 overflow and become negative.

For example
```Python
import numpy as np
np.asarray([154, 147, 151, 109], dtype=np.int8)
```
gives
```
array([-102, -109, -105,  109], dtype=int8)
```

**Comment by SabineEmbacher** [comment]
JZarr is bound to the primitive data types of Java. There is no primitive unsigned byte data type in Java.
Data types which are offered in the JZarr API (see com.bc.zarr.DataType) only serve to be written into the .zarray file so that the reader of the data knows how to interpret it. When the written data is read with Python, the interpretation happens automatically.
If the data is read again with JZarr, the correct interpretation of the raw data is up to the user.

Here is an example of interpreting byte data as unsigned bytes. [org.esa.snap.core.datamodel.ProductData.UByte](https://github.com/senbox-org/snap-engine/blob/800a6a7533a5d46b10fca7c6de075f799922de24/snap-core/src/main/java/org/esa/snap/core/datamodel/ProductData.java#L1368)

The same overflow effect will be observed with the u2 and u4 data types.

**Comment by joshmoore** [comment]
Thanks for the explanation, @SabineEmbacher. I clearly misunderstood https://jzarr.readthedocs.io/en/latest/datatype.html?highlight=unsigned#data-types

**Comment by joshmoore** [comment]
Failures now look to all be related to nesting:

<details><summary>9 failures</summary>

```
FAILED test/test_read_all.py::test_correct_read[read xtensor_zarr (nested) zarr using jzarr, blosc]
10268
FAILED test/test_read_all.py::test_correct_read[read xtensor_zarr (nested) zarr using jzarr, raw]
10269
FAILED test/test_read_all.py::test_correct_read[read xtensor_zarr (nested) zarr using jzarr, zlib]
10270
FAILED test/test_read_all.py::test_correct_read[read zarr (FSStore, nested) zarr using jzarr, blosc]
10271
FAILED test/test_read_all.py::test_correct_read[read zarr (FSStore, nested) zarr using jzarr, raw]
10272
FAILED test/test_read_all.py::test_correct_read[read zarr (FSStore, nested) zarr using jzarr, zlib]
10273
FAILED test/test_read_all.py::test_correct_read[read zarr (NestedDirectoryStore, nested) zarr using jzarr, blosc]
10274
FAILED test/test_read_all.py::test_correct_read[read zarr (NestedDirectoryStore, nested) zarr using jzarr, raw]
10275
FAILED test/test_read_all.py::test_correct_read[read zarr (NestedDirectoryStore, nested) zarr using jzarr, zlib]
```

</details>
which should be fixed once JZarr releases https://github.com/zarr-developers/zarr-python/pull/715 

@grlee77: don't know if you want to hold off on this PR until then, or introduce a way to exclude/skip/tolerate that part of the matrix.

**Comment by grlee77** [comment]
> @grlee77: don't know if you want to hold off on this PR until then, or introduce a way to exclude/skip/tolerate that part of the matrix.

I am fine with waiting, but if you would rather merge sooner we could introduce an environment variable to enable skipping the known failures.

**Comment by SabineEmbacher** [comment]
> Thanks for the explanation, @SabineEmbacher. I clearly misunderstood https://jzarr.readthedocs.io/en/latest/datatype.html?highlight=unsigned#data-types

Do you have a suggestion how I could write this in the documentation to avoid misinterpretation?

**Comment by joshmoore** [comment]
Hey @SabineEmbacher. I guess the question is if you see a way forward here. Or is this simply an incompatibility that will remain between zarr-python and jzarr?

File: generate_data/jzarr/generate_data.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# cd to this directory
+# https://stackoverflow.com/a/6393573/2700168
+cd "${0%/*}"
+
+set -e
+set -u
+set -x
+
+mvn clean package
```
Comment (grlee77) [inline generate_data/jzarr/generate_data.sh:9]:
```suggestion
mvn --no-transfer-progress clean package
```
I think adding this option will suppress the thousands of lines related to file transfer progress in the CI logs, making them more readable.

Comment (grlee77) [reply generate_data/jzarr/generate_data.sh:9]: Alternatively, or in addition, we might consider caching the `~/.m2/repository` folder on GitHub Actions (e.g. https://stackoverflow.com/a/59571393)

Comment (joshmoore) [reply generate_data/jzarr/generate_data.sh:9]: I'll add those in here, @grlee77, and then hand this off to you for a final blessing.

**Comment by joshmoore** [comment]
_Whew_. 💚  @grlee77, turning this over to you. I imagine we can continue the discussion around signedness (https://github.com/zarr-developers/zarr_implementations/pull/37#issuecomment-832194552) elsewhere.

**Review by grlee77 — APPROVED** [review]
Thanks @joshmoore, this looks good to me now.

**Pull request merged**